File: swift/plugins/Doppler.swift
// The `doppler` provider kind: Doppler.
//
// A plugin, not a built-in: it opens a socket.
//
// A port of typescript/plugins/doppler.ts, which is canonical.

import Foundation

import Sekreto

// A SCOPED import: plugin exports a `Json` of its own, and the only name
// this file wants from it is `Definition`.
import struct VoxgigPlugin.Definition

/// Doppler.
///
/// The whole config is downloaded once - Doppler's own bulk endpoint - and
/// answered from memory, like a remote .env: `api.token` is the
/// `API_TOKEN` entry. A service token is config-scoped, so project and
/// config are only needed with broader tokens.
public final class DopplerProvider: Provider {

  private let token: String?
  private let project: String?
  private let config: String?
  private let addr: String?

  private var values: Ordered<String>?

  public init(
    token: String? = nil, project: String? = nil, config: String? = nil, addr: String? = nil
  ) {
    self.token = token
    self.project = project
    self.config = config
    self.addr = addr
  }

  private func load() throws -> Ordered<String> {
    if let loaded = values { return loaded }

    let useaddr = trimslash(first(addr, "https://api.doppler.com"))
    try checkaddr(useaddr)

    var url = "\(useaddr)/v3/configs/config/secrets/download?format=json"
    if let use = project, !use.isEmpty { url += "&project=" + uriescape(use) }
    if let use = config, !use.isEmpty { url += "&config=" + uriescape(use) }

    var headers = Ordered<String>()
    headers["authorization"] = "Bearer \(token ?? "")"

    let res = try fetchjson("GET", url, headers)

    guard 200 == res.status, let body = res.body.asmap else {
      throw SekretoError("sekreto: doppler error: \(res.status)")
    }

    var loaded = Ordered<String>()

    for (key, value) in body.pairs {
      if let text = value.text { loaded[key] = text }
    }

    // Only a successful load is remembered: a failed one retries.
    values = loaded
    return loaded
  }

  public func lookup(_ name: String) throws -> String? {
    return try load()[try envkey(name)]
  }

  public func describe() -> String {
    let useproject = project ?? ""
    return useproject.isEmpty ? "doppler" : "doppler:\(useproject)/\(config ?? "")"
  }
}

/// The kind, as a voxgig/plugin definition.
public let doppler: Definition = providerplugin("doppler") { spec in
  DopplerProvider(
    token: spec.token, project: spec.project, config: spec.config, addr: spec.addr)
}

File: cpp/plugins/Doppler.cpp
#include "Doppler.hpp"

#include "Httpjson.hpp"
#include "Providers.hpp"
#include "Sekreto.hpp"

namespace sekreto {

namespace {

// ---------------------------------------------------------------- doppler

/// Doppler.
///
/// The whole config is downloaded once - Doppler's own bulk endpoint - and
/// answered from memory, like a remote .env: `api.token` is the
/// `API_TOKEN` entry. A service token is config-scoped, so project and
/// config are only needed with broader tokens.
class DopplerProvider : public Provider {
 public:
  DopplerProvider(const std::string& token, const std::string& project,
                  const std::string& config, const std::string& addr)
      : token_(token), project_(project), config_(config), addr_(addr) {}

  std::optional<std::string> lookup(const std::string& name) override {
    load();
    // The `prefix` option is deliberately not consulted by this kind.
    return values_.get(envkey(name));
  }

  std::string describe() const override {
    return project_.empty() ? "doppler" : "doppler:" + project_ + "/" + config_;
  }

 private:
  void load() {
    if (loaded_) return;

    std::string useaddr = trimslash(first(addr_, "https://api.doppler.com"));
    checkaddr(useaddr);

    std::string url = useaddr + "/v3/configs/config/secrets/download?format=json";
    if (!project_.empty()) url += "&project=" + uriescape(project_);
    if (!config_.empty()) url += "&config=" + uriescape(config_);

    Ordered headers;
    headers.set("authorization", "Bearer " + token_);

    Answer res = fetchjson("GET", url, headers);

    if (200 != res.status || !res.body.isobj()) {
      throw SekretoError("sekreto: doppler error: " + std::to_string(res.status));
    }

    Ordered values;

    for (const auto& entry : res.body.objval) {
      std::optional<std::string> text = textof(entry.second);
      if (text.has_value()) values.set(entry.first, text.value());
    }

    // Only a successful load is remembered: a failed one retries.
    values_ = values;
    loaded_ = true;
  }

  std::string token_;
  std::string project_;
  std::string config_;
  std::string addr_;

  Ordered values_;
  bool loaded_ = false;
};

}  // namespace

Definition doppler() {
  return providerplugin("doppler", [](const ProviderSpec& spec) -> std::shared_ptr<Provider> {
    return std::make_shared<DopplerProvider>(spec.token, spec.project, spec.config, spec.addr);
  });
}

}  // namespace sekreto
